File: src/main/java/cz/muni/fi/pa039/Point.java
package cz.muni.fi.pa039;

public class Point {

    public int x;
    public int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        } else if (!(other instanceof Point)) {
            return false;
        } else {
            Point p = (Point) other;
            return p.x == this.x && p.y == this.y;
        }
    }

    @Override
    public int hashCode() {
        return 37 * (this.x + 19 * this.y);
    }
}

File: src/cz/muni/fi/pa039/Point.kt
package cz.muni.fi.pa039

data class Point(var x: Int, var y: Int) {
    operator fun minus(other: Point): Point {
        return Point(x - other.x, other.y - y)
    }

    operator fun times(other: Point): Int {
        return x * other.x + y * other.y
    }

    fun vectorLength(): Double {
        return Math.sqrt(x.toDouble() * x + y * y)
    }
}